docs/flip/2434-general-metadata.md

```diff
@@ -0,0 +1,143 @@
+
+Currently, there is an experimental [API](https://github.com/google/flax/blob/main/flax/linen/partitioning.py) supporting partitioning annotations with wrappers around lifted transforms that change axes (``nn.scan_with_axes``, ``nn.vmap+with_axes``) and a special APIs to create variables (``param_with_axes`` and ``variable_with_axes``).
+The experimental partitioning API stores the metadata in a seperate collection named "[collection]_axes".
+
+
+The experimental API has a number of shortcomings that we like to solve:
+1. The current API works for tracking PartitionSpecs but not for other types of metadata like optimizer annotations.
+2. The implementation using an "xxx_axes" collection requires error-prone and non-composable string manipulation.
+3. Special, partioning-aware variable creators and lifted transforms are required
+4. The partioning API is hard to use with pre-existing Modules that aren't partioning aware.
+
+
+## Proposal
+
+To generalize metadata tracking and keep the specific metadata out of core Flax we propose the following abstract base class:
+
+```python
+class AxisMetadata(metaclass=abc.ABCMeta):
+
+  @abc.abstractmethod
+  def unbox(self) -> Any:
+    pass
+
+  @abc.abstractmethod
+  def add_axis(self, index: int, params: Dict[Any, Any]) -> TAxisMetadata:
+    pass
+
+  @abc.abstractmethod
+  def remove_axis(self, index: int, params: Dict[Any, Any]) -> TAxisMetadata:
+    pass
+```
+
+We call this type of class wrapping a value and keeping track of some additional data a **box**.
+By defining an abstract base class for this box, the API does not need to be aware of the specifics of the metadata that is tracked.
+This should make the API future proof and modular.
+
+The ``add_axis`` and ``remove_axis`` callback return an instance of their own type instead of mutating in-place.
+Typically, an implementation would be a ``flax.struct.PyTreeNode`` because the box should still be a valid JAX value and must therefore be handled by the PyTree API.
+Calling ``jax.tree_map`` on a boxed value will simply map over the value in the box.
+The lifted transforms that need to handle metadata will call ``jax.tree_map(..., is_leaf=lambda x: isinstance(x, AxisMetadata))`` to find the AxisMetadata instances within a PyTree.
+
+Advantages of the boxing approach:
+1. Boxing can be used outside of Flax and metadata is automatically "inherited". For example, the optimizerd state will
+   have the same partitioning spec as the parameters, because the state is initialized using a ``jax.tree_map`` over the boxed parameters.
+2. Boxes are composable.
+3. Boxing avoids string manipulation and generally avoids having to handle additional auxilary collections like "param_axes" in the current
+   partitioning API.
+4. No need to lift metadata collections seperately.
+
+
+Disadvantages:
```

Comment:
this also doubles the number of "levels" in the variable pytrees?
in particular, before the (un)boxing there was a 1:1 correspondence variables <-> module hierarchy
(which in turn might break some user code assumptions)

Reply:
Added